### ISSUE TITLE
[Android/Java] Overload the startUpload method with InputStream parameter

### ISSUE THREAD

**Issue by glodos** [meganz/sdk#78]
Hi,
Is it possible to add a variant of the startUpload() method which would take an InputStream as a parameter? Currently it's only possible with a local path and I have to download the file to a temporary copy, which is a nonsense. Streams are much better for data transfers and make the API more universal. 
Thanks!

**Comment by javiserrano** [comment]
I'm fully agree. Unfortunately, it's not possible with the current version of the SDK because it uses the filesystem to get the data of files.

I think that we can keep this issue open to update it when there are news in this sense.

**Comment by glodos** [comment]
Thanks, I'm hoping to see it soon.
BTW, do you know if current SDK can read from named pipes? I tried to workaround this by creating a pipe, writing the stream to it and providing its path to the startUpload method, but it always gives me a broken pipe exception. Similar approach works in other parts of my app, but not in Mega SDK.

**Comment by javiserrano** [comment]
We haven't tried to do it so, as you say, it probably doesn't work.
If you already have code that works using that approach, you could take a look at these files:
`src/posix/fs.cpp`
`src/win32/fs.cpp`

and change how the SDK accesses files (`fopen`, `sysread` and `fwrite`) to support that usage.

**Comment by tzl** [comment]
I overload startUpload() with the underlying Linux pipe(fd[2]) facility, by which the Java caller could write stream to native layer and native C MegaSDK could upload the stream content.

Details:
1: In native layer I create a Linux pipe, and get two file handles, one for MegaSDK to read and upload the data and the other for Java to write content to native layer.
2: All the rest to do are separating the read-file-handle from normal local file handlers, keeping track the read-file-handle and simulate fs.open(), fs.close(), fs.read().

In normal file uploading  mode, each process must read all the file content 3 times, 
1):reading the first time to get the file fingerprint, 2):the second time to get the file content and upload it and 3):the last time to calculate the fingerprint again to verify whether the file content is changing during the uploading. 

Now that my stream may not support seek()/open()/close() again and again, so I read the stream content only once and provide two same pseudo fingerprint before reading and after uploading.  

The  overloaded startUpload()  works but my concern is: What  does MegaSDK will do with the fingerprint besides verifying the file changing during uploading?

**Comment by javiserrano** [comment]
I think that your approach should work well, with few inconveniences.

Fingerprints are used to verify if files change during uploads, but also to check the integrity of downloads, to check if local files change during scanning in a local synced folder (only when using the sync feature) and for other optimizations of the sync algorithm.

Ideally, the fingerprint could be correctly calculated in streaming during uploads, without the need of seek() or read the data several times, but that isn't implemented yet.

You can provide a fake fingerprint in the SDK to pass the checks and ease the implementation, but please don't attach that fingerprint to the uploaded node because that would make downloads fail for users using official clients (web, mobile apps, MEGAsync, etc.).

Fingerprints are attached at the end of uploads in `src/file.cpp:92`:

`t->serializefingerprint(&attrs.map['c']);`

Commenting that line, the fingerprint wouldn't be attached to the node. It's better to not attach a fingerprint rather than attaching a wrong one.

**Comment by javiserrano** [comment]
An experimental `startUpload` function supporting input streams has been implemented in the `upload_from_inputstream` branch (but not merged yet on `master`).

I'm going close this issue and keep the other one open until this function is merged into `master`. If you have further doubts or questions about this issue, please comment them in the other one:
https://github.com/meganz/sdk/issues/120

**Comment by dknchris** [comment]
Hi, is there an actual progress on this? I don't see a startUpload method that takes InputStream. On Android, this is very critical since files on external storages cannot be accessed using raw file paths.
